lib/DiscoveryService.php

```diff
@@ -0,0 +1,192 @@
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OCA\Richdocuments;
+
+use OCP\ICache;
+use OCP\ICacheFactory;
+use OCP\ILogger;
+use OCP\Http\Client\IClientService;
+use OCA\Richdocuments\AppConfig;
+use SimpleXMLElement;
+
+class DiscoveryService {
+
+	/**
+	 * @var AppConfig
+	 */
+	private $appConfig;
+
+	/**
+	 * @var ILogger
+	 */
+	private $logger;
+
+	/**
+	 * @var ICache
+	 */
+	private $cache;
+
+	/**
+	 * @var IClientService
+	 */
+	private $httpClient;
+
+	public function __construct(
+		AppConfig $config,
+		ILogger $logger,
+		ICacheFactory $cacheFactory,
+		IClientService $httpClient
+	) {
+		$this->appConfig = $config;
+		$this->logger = $logger;
+		$this->cache = $cacheFactory->create('oca.richdocuments.discovery');
+		$this->httpClient = $httpClient;
+	}
+
+	/**
+	 * Return the original wopi url or test wopi url
+	 */
+	public function getWopiUrl(): string {
+		if ($this->appConfig->testUserSessionEnabled()) {
+			return $this->appConfig->getAppValue('test_wopi_url');
+		}
+		return $this->appConfig->getAppValue('wopi_url');
+	}
+
+	/**
+	 * Get urlsrc and action for a given mimetype from WOPI discovery
+	 *
+	 * @param string $mimetype
+	 *
+	 * @return array|null returns urlsrc and action if matched or null in case of error to retrieve discovery
+	 */
+	public function getWopiSrcUrl(string $mimetype) : ?array {
+		$discoveryXML =	$this->getDiscovery();
+		if ($discoveryXML === null) {
+			// error retrieving discovery
+			return null;
+		}
+
+		$result = $discoveryXML->xpath(\sprintf('/wopi-discovery/net-zone/app[@name=\'%s\']/action', $mimetype));
+		if (($result !== false) && (\count($result) > 0)) {
+			return [
+				'urlsrc' => (string)$result[0]['urlsrc'],
+				'action' => (string)$result[0]['name']
+			];
+		}
+
+		// no matching mimetype found
+		return [ 'urlsrc' => null, 'action' => null ];
+	}
+
+	/**
+	 * Return the content of discovery.xml - either from cache, or download it.
+	 *
+	 * @return SimpleXMLElement|null returns discovery or null in case of discovery cannot be retrieved
+	 */
+	private function getDiscovery() : ?SimpleXMLElement {
+		if ($this->appConfig->testUserSessionEnabled()) {
+			$wopiRemote = $this->appConfig->getAppValue('test_wopi_url');
+			$discoveryKey = 'discovery.xml_test';
+		} else {
+			$wopiRemote = $this->appConfig->getAppValue('wopi_url');
+			$discoveryKey = 'discovery.xml';
```

Comment:
`discoveryKey` seems a bad name. I though it was some kind of secret or password, not just a filename.